operations/agent-flow-mixin/alerts/clustering.libsonnet

```diff
@@ -25,5 +25,50 @@ alert.newGroup(
       'Cluster nodes have entered a split brain state.',
       '5m',
     ),
+
+    // Standard Deviation of Lamport clock time between nodes is too high
+    alert.newRule(
+      'ClusterLamportClockDrift',
+      'stddev by (cluster, namespace) (cluster_node_lamport_time) != 0',
+      "Cluster nodes' lamport clocks are not converging.",
+      '5m'
+    ),
```

Comment:
I think this one needs a little tweaking. Lamport clocks won't ever converge on large clusters, but you just don't want it to drift _too_ much; we don't want clocks to deviate too far from the mean clock value. 

We should start with `> 2` here, since conventionally a deviation of +/-2 is considered to be statically significant (according to [Wikipedia](https://en.wikipedia.org/wiki/Standard_deviation)). 

We should also change the description to talk more about the deviation:

```
Cluster nodes' lamport clocks time has deviated too far from the mean.
```

Reply:
I'm not sure if that expression will catch every scenario that we'd want to alert on, since the drift you're calculating is used in the calculation of stddev. 

I think if there's only one outlier, it wouldn't be alerted on since its own deviation is not double of itself. 

Consider a lamport clock set of (2, 2, 2, 2, 20). We want to alert on the 20 here. Your alert would resolve to

```
abs(20 - 5.6) > 2 * 7.2  => 
abs(14.4) > 14.4 => 
false 
```

And then if you add a second problematic node (2, 20, 2, 2, 20), nothing would be alerted on, because no node deviates more from the mean than twice the stddev value ~8.82, even though there's clearly a problem here. 

Let's take a step back:

We're trying to validate that the lamport clock is functioning properly, right? Lamport clocks must be roughly synchronized between nodes, where clocks are synchronized through sending gossiped messages. Any time a node N receives a newer clock time T, it sets its own clock time to T+1. 

I think we can pick an arbitrary value here to say that the standard deviation should be less than some target of clock ticks: clocks should always roughly be within X ticks from the mean (let's say 4 for now). If the stddev goes above 4, it means that there's a subset of nodes that are not gossiping correctly. 

What do you think?

Reply:
Yeah, let's go with this approach and see how it works in practice.